server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -325,6 +376,53 @@ public boolean rolloverOnWrite() {
         return backingIndices.rolloverOnWrite;
     }
 
+    public ComposableIndexTemplate getEffectiveIndexTemplate(ProjectMetadata projectMetadata) {
+        return mergeSettingsIntoTemplate(getMatchingIndexTemplate(projectMetadata), settings);
+    }
+
+    public Settings getEffectiveSettings(ProjectMetadata projectMetadata) {
+        ComposableIndexTemplate template = getMatchingIndexTemplate(projectMetadata);
+        return mergeSettings(template.template() == null ? Settings.EMPTY : template.template().settings(), settings);
+    }
+
+    private ComposableIndexTemplate getMatchingIndexTemplate(ProjectMetadata projectMetadata) {
+        return lookupTemplateForDataStream(name, projectMetadata);
+    }
+
+    public static ComposableIndexTemplate mergeSettingsIntoTemplate(ComposableIndexTemplate template, Settings settings) {
+        if (Settings.EMPTY.equals(settings)) {
+            return template;
+        }
+        ComposableIndexTemplate.Builder mergedIndexTemplateBuilder = template.toBuilder();
+        Template.Builder mergedTemplateBuilder;
+        Settings templateSettings;
+        if (template.template() == null) {
+            mergedTemplateBuilder = Template.builder();
+            templateSettings = null;
+        } else {
+            mergedTemplateBuilder = Template.builder(template.template());
+            templateSettings = template.template().settings();
+        }
+        mergedTemplateBuilder.settings(mergeSettings(templateSettings, settings));
+        mergedIndexTemplateBuilder.template(mergedTemplateBuilder);
+        return mergedIndexTemplateBuilder.build();
+    }
+
+    private static Settings mergeSettings(Settings originalSettings, Settings newSettings) {
+        if (newSettings == null || Settings.EMPTY.equals(newSettings)) {
+            return Objects.requireNonNullElse(originalSettings, Settings.EMPTY);
+        } else if (originalSettings == null || Settings.EMPTY.equals(originalSettings)) {
+            return newSettings;
+        }
+        Settings.Builder settingsBuilder = Settings.builder().put(originalSettings).put(newSettings);
+        for (String settingName : new HashSet<>(settingsBuilder.keys())) {
+            if (settingsBuilder.get(settingName) == null) {
+                settingsBuilder.remove(settingName);
+            }
+        }
+        return settingsBuilder.build();
+    }
```

Comment:
I'm not sure but this feels like the wrong place for this kind of method... Could the this and mergeSettingsIntoTemplate be part of the template builder or Settings object for better re-use / encapsulation?

Reply:
Yeah I agree. It makes me a little nervous to move them because now their behavior is not clearly just meant for data streams. See what you think of the latest commit.

Reply:
I think I do prefer it with the new layout... Is there a risk if things other than datastreams use it?